doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,1208 @@
+and extensible interface.
+The end-user will not need to be aware of this.
+
+.. _hierarchy_figure:
+.. figure:: _static/dtype_hierarchy.svg
+    :figclass: align-center
+    
+
+Methods/Slots defined for each DType
+""""""""""""""""""""""""""""""""""""
+
+NEP 41 detailed that all logic should be defined through special methods
+on the DTypes.
+This section will list a specific set of such methods (in the form of 
+Python methods).
+The C-side equivalent slot signature will be summarized below after proposing
+the general C-API for defining new Datatypes.
+Note that while the slots are defined as special Python methods here, this is
+for convenience to the reader and *not* meant to imply the identical exposure
+as a Python API.
+This will need to be proposed in a separate, later, NEP.
+
+Some of the methods may be similar or even reuse existing Python slots.
+User-defined DType classes are discouraged from defining or using Python's
+special slots without feedback, in order to allow defining them later.
+For example ``dtype1 & dtype2`` (or maybe ``|``) could be a shorthand for
+``np.common_dtype``, and comparisons should be defined mainly through casting
+logic.
+
+
+Additional Information
+^^^^^^^^^^^^^^^^^^^^^^
+
+In addition to the more detailed methods below, the following general
+information is currently provided and will be defined on the class:
+
+* ``cls.parametric`` (see also `NEP 40 <NEP40>`_):
+
+  * This flag is mainly to simplify DType and casting creation and
+    allow for performance tweaks.
+  * DTypes which are not parametric must define a canonical dtype instance
+    which should be a singleton.
+  * Parametric dtypes require some additional methods (below).
+  * One can alternatively create a ``ParametricDType`` class for this purpose,
+    although for the C-side API a flag seems simpler.
+
+* ``self.canonical`` method (Alternative: new instance attribute)
+
+  * Instead of byteorder, we may want an ``canonical`` flag (reusing the
+    ISNBO flag – "is native byte order" seems possible here).
+    This flag signals that the data are stored in the default/canonical way.
+    In practice this is always an NBO check, but generalization should be possible.
+    A potential use case is a complex-conjugated instance of Complex which
+    stores ``real - imag`` instead of ``real + imag`` and is thus not
+    the canonical storage.
+
+* ``ensure_canonical(self) -> dtype`` return a new dtype (or ``self``).
+   The returned dtype must have the ``canonical`` flag set.
+
+* ``DType.type`` is the associated scalar type.  ``dtype.type`` will be a
+  class attribute and the current ``dtype.type`` field considered
+  deprecated. This may be relaxed if a usecase arises.
+
+Additionally, existing methods (and C-side fields) will be provided, although
+fields such as the "kind" and and "char" may be set to invalid values on
+the C-side, and access may error on the Python side.
```

Comment:
Hmm, very good point. I thought it might error for new dtypes. But you are right, it would be better if it does not.  This means that `"kind"` must return *something* though.

There are two options:

1. The dtype author has to provide a `kind`
2. We simply return the same kind (e.g. None) when it is not defined.

The only problem I see is code which for some reason would do `dtype1.kind == dtype2.kind`, I suppose this is rare, since most will be `dtype.kind == "i"`, etc.  Also on the C-side we are limited to a single struct.  For now I thought we can simply return `\0` for all user dtypes.

In the end, this would mean user dtypes must use `isinstance` checks or similar instead in the long run.

@jorisvandenbossche do you have an opinion on this or does pandas have prior art?

Reply:
It's maybe not a problem that it is undefined (or None) for user dtypes, I was mainly thinking about the built-in types like int64. 
In pandas we eg use `kind` as a fast check for certain dtypes (like `dtype.kind in ("u", "i")` for integer dtypes). 

For our own dtypes, we define a "kind" on the dtype object (for compatibility I assume), but it is not always clear what this should be or how it is useful. I think we mostly took the the numpy one if there is a direct equivalent (eg our masked integer dtype also gives kind="i"/"u"), or otherwise just "O" for things like categorical and interval dtypes.

Reply:
Maybe the default should be to return the `dtype` itself (or perhaps its name!?), so it is guaranteed to compare equal only to itself?